File: src/js/asynchrony/07_semaphore.js
/*
This script demonstrates the use of a semaphore to limit concurrent access to a 
resource in an asynchronous environment. A semaphore allows a specified number 
of tasks to access a resource simultaneously, while other tasks must wait until 
a slot becomes available.

In this example, we create a simple Semaphore class and use it to limit 
concurrent access to a simulated resource to a maximum of 3 tasks at a time.
*/

class Semaphore {
  constructor(maxConcurrent) {
    this.maxConcurrent = maxConcurrent;
    this.current = 0;
    this.queue = [];
  }

  async acquire() {
    if (this.current < this.maxConcurrent) {
      this.current++;
      return;
    }

    return new Promise((resolve) => {
      this.queue.push(resolve);
    });
  }

  release() {
    if (this.queue.length > 0) {
      const resolve = this.queue.shift();
      resolve();
    } else {
      this.current--;
    }
  }
}

const resourceSemaphore = new Semaphore(3);

async function limitedResource(taskId) {
  await resourceSemaphore.acquire();
  console.log(`Task ${taskId} is using the limited resource.`);
  await new Promise((resolve) =>
    setTimeout(resolve, Math.floor(Math.random() * 2000) + 500)
  );
  console.log(`Task ${taskId} is done using the limited resource.`);
  resourceSemaphore.release();
}

async function main() {
  const tasks = [];

  for (let i = 0; i < 10; i++) {
    tasks.push(limitedResource(i));
  }

  await Promise.all(tasks);
}

main();

File: src/python/asynchrony/07_semaphore.py
"""
This Python script demonstrates the use of an async semaphore to control access
to a limited resource in an asynchronous context. Async semaphores work similarly
to threading semaphores but are specifically designed for use with async functions
and coroutines.

In this example, we simulate multiple tasks attempting to access a limited resource
simultaneously. The resource can only handle three concurrent accesses. We use
an async semaphore to ensure that no more than three tasks access the resource at
the same time.
"""

import asyncio
import random

# The limited resource can only handle three concurrent accesses.
resource_semaphore = asyncio.Semaphore(3)


async def limited_resource(task_id):
    async with resource_semaphore:
        print(f"Task {task_id} is using the limited resource.")
        await asyncio.sleep(random.uniform(0.5, 2))
        print(f"Task {task_id} is done using the limited resource.")


async def main():
    tasks = [asyncio.create_task(limited_resource(i)) for i in range(10)]
    await asyncio.gather(*tasks)


if __name__ == "__main__":
    asyncio.run(main())
